### PR TITLE
combine keep_attrs and combine_attrs in apply_ufunc

### PR DESCRIPTION
**Pull request by keewis**
As discussed this extends `keep_attrs` in `apply_ufunc` to also accept merge strategy names and use `merge_attrs` (making it easier to add more merge strategies).

Handling `attrs` on variables / coords is still missing.

- [x] Tests added
- [x] Passes `pre-commit run --all-files`
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

### CODE REVIEW

**Comment by keewis** [comment]
it seems `apply_ufunc` already applies `keep_attrs`/`combine_attrs` on the data variables, but not the coords. Not sure if we should try to combine attrs on coords, too (should be fixed by adding a `keep_attrs` kwarg to `align` / `deep_align`)

File: xarray/tests/test_computation.py [hunk 1]
```diff
@@ -563,6 +564,361 @@ def add(a, b, keep_attrs):
...
+        ),
+        pytest.param(
+            "no_conflicts",
+            [{"a": 1}, {"b": 2}, {"b": 3}],
+            None,
+            True,
+            id="no_conflicts",
+        ),
+    ),
+)
```
Comment (max-sixty) [inline xarray/tests/test_computation.py:620]: These are such clear test cases, very nice!

**Review by max-sixty — APPROVED** [review]
No strong view on how we combine on the coords. Though is there any reason to wait on merging this? The tests are great!

**Comment by dcherian** [comment]
> if we should try to combine attrs on coord

:+1: on this. can fix later though...

**Comment by keewis** [comment]
`apply_ufunc` ultimately calls `merge_collected` to determine the output coords so we would need #4902 to also keep the attrs on coords.

**Comment by keewis** [comment]
this is ready, too. Not sure if we should merge before the release, though.

**Comment by max-sixty** [comment]
> this is ready, too. Not sure if we should merge before the release, though.

As good a time as any! But as ever your call on whether we can be sufficiently confident

File: xarray/core/computation.py [hunk 2]
```diff
@@ -801,7 +822,7 @@ def apply_ufunc(
     join: str = "exact",
     dataset_join: str = "exact",
     dataset_fill_value: object = _NO_FILL_VALUE,
-    keep_attrs: bool = False,
+    keep_attrs: Union[bool, str, None] = False,
```
Comment (keewis) [inline xarray/core/computation.py:825]: should the default value become `None` (falling back to `False` in the `_get_keep_attrs` call)?

**Review by keewis — COMMENTED** [review]
considering that we originally planned to release tomorrow (?) I'd say let's wait and include this in the next release. I just went through the code and found a few issues which need more thought (I think?)

**Comment by keewis** [comment]
not sure if I got all of the issues I thought I noticed, but I think this should be ready

**Review by max-sixty — APPROVED** [review]
Tests look good!

(Whatsnew needs switching versions FYI)

**Comment by keewis** [comment]
> (Whatsnew needs switching versions FYI)

thanks, I completely missed that

**Comment by dcherian** [comment]
Thanks @keewis. I found the `compute_attrs` thing a little complicated and hard to understand. But I think it's OK and am going to merge now. If you have a clever idea for something clearer, we can address that later.

Thanks!

**Pull request merged**